### PR TITLE
 Added ProtoSize wrapper functions for the well known types  #438

### PR DESCRIPTION
**Pull request by jmarais**
@awalterschulze 
Is there a way to generate the Value_** types? I tried, but failed.

### CODE REVIEW

**Comment by awalterschulze** [comment]
@jmarais 
Interesting, they seem to have generated in wrappers.pb.go
https://github.com/gogo/protobuf/blob/master/types/wrappers.pb.go

Here they are used
https://github.com/gogo/protobuf/blob/ebc05652de1e3bd0da5f0879f2548d6580eee8fa/jsonpb/jsonpb_test_proto/test_objects.proto#L150

So this makefile should show you have to generate them
https://github.com/gogo/protobuf/blob/ebc05652de1e3bd0da5f0879f2548d6580eee8fa/jsonpb/jsonpb_test_proto/Makefile

What is the error message you are getting?

**Comment by jmarais** [comment]
It seems like most of the "Values" are fine. Only NumberValue and StructValue gives a problem:

```bash
issue438.proto:34:3: "google.protobuf.NumberValue" is not defined.
issue438.proto:37:3: "google.protobuf.StructValue" is not defined.
```
I see that they are not defined in wrappers.pb.go

**Comment by jmarais** [comment]
Ok. So I can change the StructValue to Struct. And leave out the NumberValue.

**Comment by jmarais** [comment]
Hmm. I think we are generating the NullValue wrong.

```golang
func (m *Types) GetNlval() types.NullValue {
	if m != nil {
		return m.Nlval
	}
	return types.NullValue_NULL_VALUE
}
```

```bash
./issue438.pb.go:157:9: undefined: types.NullValue_NULL_VALUE
```

There is only this const defined in struct.pb.go:

```golang
const (
	// Null value.
	NULL_VALUE NullValue = 0
)
```

**Comment by jmarais** [comment]
It is this line causing the wrong enum name:
https://github.com/gogo/protobuf/blob/master/protoc-gen-gogo/generator/generator.go#L2686

**Comment by awalterschulze** [comment]
Good catch

So maybe we should remove this line?
https://github.com/gogo/protobuf/blob/master/protoc-gen-gogotypes/main.go#L47

**Comment by jmarais** [comment]
That seems to work.
It generates:
```golang
const (
	// Null value.
	NullValue_NULL_VALUE NullValue = 0
)
```
in struct.pb.go.

Do we want to make that change? I am unsure what else it will affect (if anything).

**Comment by awalterschulze** [comment]
I think lets make the change and see the diff :)

What do you think?

File: types/type.pb.go [hunk 1]
```diff
@@ -32,9 +32,9 @@ type Syntax int32
 
 const (
 	// Syntax `proto2`.
-	SYNTAX_PROTO2 Syntax = 0
+	Syntax_SYNTAX_PROTO2 Syntax = 0
```
Comment (awalterschulze) [inline types/type.pb.go]: Here is the one other change

File: types/type.pb.go [hunk 2]
```diff
@@ -47,51 +47,51 @@ var Syntax_value = map[string]int32{
...
+	return fileDescriptor_type_0082d870c49329d7, []int{0}
 }
 
 // Basic field types.
 type Field_Kind int32
 
 const (
 	// Field type unknown.
-	TYPE_UNKNOWN Field_Kind = 0
+	Field_TYPE_UNKNOWN Field_Kind = 0
```
Comment (awalterschulze) [inline types/type.pb.go]: Oh and here

File: types/type.pb.go [hunk 3]
```diff
@@ -138,21 +138,21 @@ var Field_Kind_value = map[string]int32{
...
+	return fileDescriptor_type_0082d870c49329d7, []int{1, 0}
 }
 
 // Whether a field is optional, required, or repeated.
 type Field_Cardinality int32
 
 const (
 	// For fields with unknown cardinality.
-	CARDINALITY_UNKNOWN Field_Cardinality = 0
+	Field_CARDINALITY_UNKNOWN Field_Cardinality = 0
```
Comment (awalterschulze) [inline types/type.pb.go]: more

**Comment by awalterschulze** [comment]
I think given those are the only changes, then it is ok to generate the enum prefixes.

**Comment by awalterschulze** [comment]
@johanbrandhorst are these enum prefixes for well known types affecting anything in grpc?

**Comment by jmarais** [comment]
Yeh. Currently those values are only used in the /types package.

**Comment by johanbrandhorst** [comment]
:grimacing: I'd hate to think anything hardcodes the enum values, I think all enums are looked up from the proto registry.

**Comment by awalterschulze** [comment]
Ok then I think we are ready to merge :)
@jmarais are you ready?

**Comment by jmarais** [comment]
Yup. Sounds good.

**Pull request merged**